torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,188 @@
+    its triton generated hash, its device, and where this kernel is located.
+    This is the minimum information we can use to later retreive this kernel
+    from file system.
+    """
+
+    kernel_hash: str
+    compile_time_ns: int
+    device: int
+    directory: str
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonKernelArtifact:
+    """
+    Artifact for an individual kernel converted to bytes.
+    Bytes could be a cubin, json, ttir, or ttgir.
+    """
+
+    filename: str
+    payload: bytes = dataclasses.field(repr=False)  # Do not display binary
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonKernelArtifacts:
+    """
+    Collection of artifacts for a particular kernel.
+    """
+
+    kernel_hash: str
+    compile_time_ns: int
+    device: int
+    artifacts: List[TritonKernelArtifact]
+
+
+@dataclasses.dataclass(frozen=True)
+class TritonBundlerMetadata:
+    """
+    Metadata used for instrumentation
+    """
+
+    num_kernels: int
+    total_compile_time_ns: int
+
+
+class TritonBundler:
+    """
+    Lightweight Triton Kernel bundler that notes each time we compile a triton
+    kernel. When collect is called, converts all the previously noted kernels and
+    their artifacts into a structured bytes blob, and later when write is called
+    it writes this structured blob back to file system.
+    """
+
+    # It is safe for multiple threads to insert entries as insert operation
+    # of OrderedSet which uses dict underhood is atomic
+    _entries: Optional[OrderedSet[TritonBundleEntry]] = None
+
+    @staticmethod
+    def is_enabled() -> bool:
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
+
+        if not config.is_fbcode():
+            return False
+
+        return justknobs_check("pytorch/remote_cache:bundle_triton_into_fx_graph_cache")
+
+    @classmethod
+    def begin_compile(cls) -> None:
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = OrderedSet()
+
+    @classmethod
+    def put(cls, kernel_hash: str, compile_time_ns: int, device: int) -> None:
+        if (entries := cls._entries) is not None:
+            # CachingAutotuner.__init__ unconditionally sets TRITON_CACHE_DIR
+            triton_cache_dir = os.getenv("TRITON_CACHE_DIR")
+            assert triton_cache_dir is not None
+            entries.add(
+                TritonBundleEntry(
+                    kernel_hash, compile_time_ns, device, triton_cache_dir
+                )
+            )
+
+    @classmethod
+    def collect(
+        cls,
+    ) -> Tuple[List[TritonKernelArtifacts], Optional[TritonBundlerMetadata]]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
+            return [], None
+
+        with dynamo_timed(
+            key="TritonBundler::collect", phase_name="TritonBundler", fwd_only=False
```

Comment:
why colons in the key?

Reply:
@jamesjwu I'm not convinced these should have phase name. It's good to have these available in the detailed table but it's not clear I want two new dedicated columns in dynamo_compile for these timings only. Especially because these are filesystem interaction only.